File: palindrome-linked-list/solution234.go
// https://leetcode.com/problems/palindrome-linked-list/

package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func NewListNode(val int) *ListNode {
	return &ListNode{Val: val, Next: nil}
}

// T: O(n), S: O(1)
func isPalindrome(head *ListNode) bool {
	firstHalfEnd := firstHalfEnd(head)
	secondHalfStart := reverseList(firstHalfEnd.Next)

	result := true
	for curr1, curr2 := head, secondHalfStart; result && curr2 != nil; curr1, curr2 = curr1.Next, curr2.Next {
		if curr1.Val != curr2.Val {
			result = false
		}
	}
	firstHalfEnd.Next = reverseList(secondHalfStart)
	return result
}

func firstHalfEnd(head *ListNode) *ListNode {
	slow, fast := head, head
	for fast.Next != nil && fast.Next.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	return slow
}

func reverseList(node *ListNode) *ListNode {
	var prev *ListNode
	curr := node
	for curr != nil {
		tempNext := curr.Next
		curr.Next = prev
		prev = curr
		curr = tempNext
	}
	return prev
}

// T: O(n), S: O(n)
func isPalindromeHalfStack(head *ListNode) bool {
	var st []int
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		st = append(st, slow.Val)
		slow = slow.Next
		fast = fast.Next.Next
	}
	if fast != nil {
		slow = slow.Next
	}
	i := len(st) - 1
	for slow != nil && len(st) != 0 {
		val := slow.Val
		if val != st[i] {
			return false
		}
		i--
		slow = slow.Next
	}
	return true
}

// T: O(n), S: O(n)
func isPalindromeFullStack(head *ListNode) bool {
	var st []int
	curr := head
	for curr != nil {
		st = append(st, curr.Val)
		curr = curr.Next
	}
	curr = head
	i := len(st) - 1
	for curr != nil {
		val := st[i]
		if val != curr.Val {
			return false
		}
		i--
		curr = curr.Next
	}
	return true
}

func main() {
	// [1,2,2,1]
	head1 := NewListNode(1)
	head1.Next = NewListNode(2)
	head1.Next.Next = NewListNode(2)
	head1.Next.Next.Next = NewListNode(1)
	fmt.Println(isPalindrome(head1))          // true
	fmt.Println(isPalindromeHalfStack(head1)) // true
	fmt.Println(isPalindromeFullStack(head1)) // true
	// [1,2]
	head2 := NewListNode(1)
	head2.Next = NewListNode(2)
	fmt.Println(isPalindrome(head2))          // false
	fmt.Println(isPalindromeHalfStack(head2)) // false
	fmt.Println(isPalindromeFullStack(head2)) // false
	// [1,1,2,1]
	head3 := NewListNode(1)
	head3.Next = NewListNode(1)
	head3.Next.Next = NewListNode(2)
	head3.Next.Next.Next = NewListNode(1)
	fmt.Println(isPalindrome(head3))          // false
	fmt.Println(isPalindromeHalfStack(head3)) // false
	fmt.Println(isPalindromeFullStack(head3)) // false
	// [1,0,1]
	var head4 = NewListNode(1)
	head4.Next = NewListNode(0)
	head4.Next.Next = NewListNode(1)
	fmt.Println(isPalindrome(head4))          // true
	fmt.Println(isPalindromeHalfStack(head4)) // true
	fmt.Println(isPalindromeFullStack(head4)) // true
}

File: palindrome-linked-list/solution234.py
# https://leetcode.com/problems/palindrome-linked-list/

from typing import Optional


class ListNode:

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution234:

    # T: O(n), S: O(n)
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        lst = []
        curr = head
        while curr:
            lst.append(curr.val)
            curr = curr.next
        left, right = 0, len(lst) - 1
        while left < right:
            if lst[left] != lst[right]:
                return False
            left += 1
            right -= 1
        return True

    # T: O(n), S: O(1)
    def isPalindromeInPlace(self, head: Optional[ListNode]) -> bool:
        first_half_end = self.firstHalfEnd(head)
        second_half_start = self.reverseList(first_half_end.next)
        curr1, curr2 = head, second_half_start
        result = True
        while result and curr2:
            if curr1.val != curr2.val:
                result = False
            curr1 = curr1.next
            curr2 = curr2.next
        first_half_end.next = self.reverseList(second_half_start)
        return result

    def firstHalfEnd(self, head: ListNode) -> ListNode:
        slow, fast = head, head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        return slow

    def reverseList(self, node: ListNode) -> ListNode:
        prev, curr = None, node
        while curr:
            temp_next = curr.next
            curr.next = prev
            prev = curr
            curr = temp_next
        return prev

    # T: O(n), S: O(n)
    def isPalindromeHalfStack(self, head: Optional[ListNode]) -> bool:
        st = []
        slow, fast = head, head
        while fast and fast.next:
            st.append(slow.val)
            slow = slow.next
            fast = fast.next.next
        if fast:
            slow = slow.next
        while st and slow:
            val = st.pop()
            if val != slow.val:
                return False
            slow = slow.next
        return True

    # T: O(n), S: O(n)
    def isPalindromeFullStack(self, head: Optional[ListNode]) -> bool:
        st = []
        curr = head
        while curr:
            st.append(curr.val)
            curr = curr.next
        curr = head
        while curr:
            val = st.pop()
            if curr.val != val:
                return False
            curr = curr.next
        return True


def main() -> None:
    sol = Solution234()
    # [1,2,2,1]
    head1 = ListNode(1)
    head1.next = ListNode(2)
    head1.next.next = ListNode(2)
    head1.next.next.next = ListNode(1)
    print(sol.isPalindrome(head1))  # True
    print(sol.isPalindromeInPlace(head1))  # True
    print(sol.isPalindromeHalfStack(head1))  # True
    print(sol.isPalindromeFullStack(head1))  # True
    # [1,2]
    head2 = ListNode(1)
    head2.next = ListNode(2)
    print(sol.isPalindrome(head2))  # False
    print(sol.isPalindromeInPlace(head2))  # False
    print(sol.isPalindromeHalfStack(head2))  # False
    print(sol.isPalindromeFullStack(head2))  # False
    # [1,1,2,1]
    head3 = ListNode(1)
    head3.next = ListNode(1)
    head3.next.next = ListNode(2)
    head3.next.next.next = ListNode(1)
    print(sol.isPalindrome(head3))  # False
    print(sol.isPalindromeInPlace(head3))  # False
    print(sol.isPalindromeHalfStack(head3))  # False
    print(sol.isPalindromeFullStack(head3))  # False
    # [1,0,1]
    head4 = ListNode(1)
    head4.next = ListNode(0)
    head4.next.next = ListNode(1)
    print(sol.isPalindrome(head4))  # True
    print(sol.isPalindromeInPlace(head4))  # True
    print(sol.isPalindromeHalfStack(head4))  # True
    print(sol.isPalindromeFullStack(head4))  # True


if __name__ == "__main__":
    main()
